cloudflare/resource_cloudflare_waiting_room.go

```diff
@@ -0,0 +1,223 @@
+package cloudflare
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	cloudflare "github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceCloudflareWaitingRoom() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudflareWaitingRoomCreate,
+		Read:   resourceCloudflareWaitingRoomRead,
+		Update: resourceCloudflareWaitingRoomUpdate,
+		Delete: resourceCloudflareWaitingRoomDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceCloudflareWaitingRoomImport,
+		},
+
+		SchemaVersion: 0,
+		Schema: map[string]*schema.Schema{
+			"zone_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				StateFunc: func(i interface{}) string {
+					return strings.ToLower(i.(string))
+				},
+			},
+
+			"host": {
+				Type:     schema.TypeString,
+				Required: true,
+				StateFunc: func(i interface{}) string {
+					return strings.ToLower(i.(string))
+				},
+			},
+
+			"path": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"total_active_users": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"new_users_per_minute": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"custom_page_html": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"queue_all": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+
+			"disable_session_renewal": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+
+			"suspended": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"session_duration": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Second),
+			Update: schema.DefaultTimeout(30 * time.Second),
+		},
+	}
+}
+
+func buildWaitingRoom(d *schema.ResourceData) cloudflare.WaitingRoom {
+	return cloudflare.WaitingRoom{
+		Name:                  d.Get("name").(string),
+		Description:           d.Get("description").(string),
+		Suspended:             d.Get("suspended").(bool),
+		Host:                  d.Get("host").(string),
+		Path:                  d.Get("path").(string),
+		TotalActiveUsers:      d.Get("total_active_users").(int),
+		NewUsersPerMinute:     d.Get("new_users_per_minute").(int),
+		CustomPageHtml:        d.Get("custom_page_html").(string),
+		SessionDuration:       d.Get("session_duration").(int),
+		QueueAll:              d.Get("queue_all").(bool),
+		DisableSessionRenewal: d.Get("disable_session_renewal").(bool),
+	}
+}
+
+func resourceCloudflareWaitingRoomReadFromWaitingRoomId(d *schema.ResourceData, meta interface{}, zoneID, waitingRoomID string) error {
```

Comment:
why is this a seperate method? `d.Set` should really on be called in the `Read` method which in turn is called by the other CRUD methods.

it looks like you're calling this in the `Import` method as well which we shouldn't be doing but instead relying on the `Read` method to do it's thing.

Reply:
Done